## Data Wrangling & Subsetting

In [1]:
import pandas as pd
import numpy as np
import os

In [41]:
path = r'C:\Users\efens\cf_tasks\2023-07 Instacard Basket Analysis'

# Original data Products
df_prods = pd.read_csv(os.path.join(path, '02 Data', '021 Original Data', 'products.csv'), index_col = False)

# Changed data
df_ords = pd.read_csv(os.path.join(path, '02 Data', '022 Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_deps_t = pd.read_csv(os.path.join(path, '02 Data', '022 Prepared Data', 'departments_transposed.csv'), index_col = False)

In [3]:
## I used changes csv's from the exercised

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [6]:
df_deps_t.head()

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

## 01. Identifier variable in the df_ords dataframe

In [8]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [10]:
## User_Id to string

df_ords['user_id'] = df_ords['user_id'].astype('str')
df_ords['user_id'].dtype

dtype('O')

## 02. Looking for a variable in df_ords dataframe with an unintuitive name and changing its name without overwriting the data frame.

In [11]:
## Renaming the column

df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

## 03. What the busiest hour is for placing orders?

In [12]:
# calculating the frequency

df_ords['order_hour_of_day'].value_counts(dropna = True)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

#### Answer: 10 hours is a busiest hour for placing orders (has the highest count = 288418)

## 04. The meaning behind a value of 4 in the "department_id" column within the df_prods dataframe

In [72]:
## Creating a dictionary
data_dict = df_deps_t.to_dict('records')
data_dict

[{'department_id': 1, 'department': 'frozen'},
 {'department_id': 2, 'department': 'other'},
 {'department_id': 3, 'department': 'bakery'},
 {'department_id': 4, 'department': 'produce'},
 {'department_id': 5, 'department': 'alcohol'},
 {'department_id': 6, 'department': 'international'},
 {'department_id': 7, 'department': 'beverages'},
 {'department_id': 8, 'department': 'pets'},
 {'department_id': 9, 'department': 'dry goods pasta'},
 {'department_id': 10, 'department': 'bulk'},
 {'department_id': 11, 'department': 'personal care'},
 {'department_id': 12, 'department': 'meat seafood'},
 {'department_id': 13, 'department': 'pantry'},
 {'department_id': 14, 'department': 'breakfast'},
 {'department_id': 15, 'department': 'canned goods'},
 {'department_id': 16, 'department': 'dairy eggs'},
 {'department_id': 17, 'department': 'household'},
 {'department_id': 18, 'department': 'babies'},
 {'department_id': 19, 'department': 'snacks'},
 {'department_id': 20, 'department': 'deli'},
 {'dep

In [76]:
# check what stands behind 4 using a dictionary
# print(data_dict.get('4'))
# can't get the answer with the code from the lecture, so I tried to find another way

print(data_dict[3])

{'department_id': 4, 'department': 'produce'}


In [77]:
## One more way

# define the new variable with the value 4
department_id_to_find = 4

# define the variable with the name of the department
# use the function next() to return the first matching value
# chect if the 'department_id' in that dictionary matches the value of department_id_to_find (4)
# return "None" is nothing is found
department_info = next((item for item in data_dict if item['department_id'] == department_id_to_find), None)

# Print output
print(department_info)

{'department_id': 4, 'department': 'produce'}


# 05. Breakfast sales

In [78]:
## Create a subset containing only the required information > "Breakfast"

df_breakfast = df_prods.loc[df_prods['department_id']==14]
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# 06. All observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood.

In [83]:
## Create a subset containing only the required information

df_dinner = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# 07. How many rows does the last dataframe you created have?

In [84]:
num_rows = df_dinner.shape[0]
print("Number of rows in df_dinner:", num_rows)

Number of rows in df_dinner: 7650


# 08. Information about user_id = 1

In [86]:
user_id_1 = df_ords.loc[df_ords['user_id']==1]
user_id_1

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   1187899        1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

In [91]:
# using readable values format
pd.options.display.float_format = '{:.2f}'.format
# convert oder_id to string
df_ords['order_id'] = df_ords['order_id'].astype(str)
# statistics about user_id = 1
user_id_1.describe()

order_id  user_id  order_number  orders_day_of_week  \
count      11.00    11.00         11.00               11.00   
mean  1923450.09     1.00          6.00                2.64   
std   1071950.28     0.00          3.32                1.29   
min    431534.00     1.00          1.00                1.00   
25%    869017.00     1.00          3.50                1.50   
50%   2295261.00     1.00          6.00                3.00   
75%   2544845.50     1.00          8.50                4.00   
max   3367565.00     1.00         11.00                4.00   

       order_hour_of_day  days_since_prior_order  
count              11.00                   10.00  
mean               10.09                   19.00  
std                 3.48                    9.03  
min                 7.00                    0.00  
25%                 7.50                   14.25  
50%                 8.00                   19.50  
75%                13.00                   26.25  
max                16.00                   30.00

#  09. What basic stats can you provide based on the information you have?

In [94]:
#### Answer: 
# The customer with user_id = 1 has made 11 orders, with most orders placed at the beginning of the week on Tuesdays and Wednesdays. 
# The customer tends to place orders around 11 AM, with the earliest order placed at 7 AM and the latest order at 4 PM. 
# Average time between orders is 19 days.